### PR TITLE
fix multiRead and add getHandles functionality

### PR DESCRIPTION
**Pull request by ovi1337**
This PR will implement the multiRead successfully by using the exactly the same like read or write commands.
Maybe we can simplify the API later, that you can also use an array in the read/write command and the lib is internally using the multiRead. (multiWrite should also be implemented then)

### CODE REVIEW

**Comment by ovi1337** [comment]
@ChrisHanuta have you maybe an idea about this issue?

I've also created a topic about this issue in a german SPS community, but there's no response currently:
https://www.sps-forum.de/codesys-und-iec61131/93118-mit-der-nutzung-von-handles-im-adsigrp_sumup_readwrite-request.html#post699030

**Comment by ovi1337** [comment]
Currently there's no other way to do a multiRead without IndexGroup and IndexOffset, but in general this fix is working well and as expected. I will remove the fallback to use this method with `symbolNames` only, add an error message, for prevent the wrong usage and would like to merge it, that we can close the open tickets and have a clear way to use the multiRead.

What are you guys thinking about it?

**Comment by ChrisHanuta** [comment]
Did you look at that?
https://infosys.beckhoff.com/index.php?content=../content/1031/tcadscommon/html/tcadscommon_introads.htm

It just looks like more than a Handel

**Comment by ChrisHanuta** [comment]
Sorry wrong link: https://infosys.beckhoff.com/content/1031/tcsample_vc/html/tcadsdll_api_cpp_sample18.htm?id=7560953427621100386

**Comment by ovi1337** [comment]
Mmh, it seems to be this what i've already implemented here: https://github.com/roccomuso/node-ads/pull/12/files#diff-94814b9df475840da8a076fabd72c51dR275

The problem is, that i don't know how i can use the handles in the multiRead command.
It must be something similar to https://github.com/roccomuso/node-ads/blob/master/lib/ads.js#L354
But currently i don't have figured out the way to use it in the multiRead.

I'm pretty sure that i also must use 
```
indexGroup: 0xF005
indexOffset: <symHandle>
```
But then i just get `state: 1794` as callback.
Maybe you have a good idea?

**Comment by ChrisHanuta** [comment]
This is the getMultiHandle:

var getMultiHandle = function (handles, cb) {
  var ads = this
  var bufs = null
  var bufsymnames = null
  var countsymnames = 0
  handles.forEach( function (handle){
    handle = parseHandle(handle)
    if (typeof handle.symname === 'undefined') {
      handle.symname = handle.indexOffset
      cb.call(ads, null, handle)
    } else {
      var bufsymname = stringToBuffer(handle.symname)
      if (bufsymnames) {
        bufsymnames = Buffer.concat([bufsymnames, bufsymname ], bufsymnames.length + bufsymname.length)
      } else {
        bufsymnames = bufsymname
      }
      var buf = Buffer.alloc(16)
      buf.writeUInt32LE(ADSIGRP.GET_SYMHANDLE_BYNAME,0)
      buf.writeUInt32LE(0x00000000,4)
      buf.writeUInt32LE(4,8)
      buf.writeUInt32LE(bufsymname.length,12)
      if (bufs) {
        bufs = Buffer.concat([bufs, buf ], bufs.length + buf.length)
      } else {
        bufs = buf
      }
      countsymnames++
    }
  });
  
  if (bufs && bufsymnames){
    var buf = Buffer.concat([bufs, bufsymnames ], bufs.length + bufsymnames.length)
    var commandOptions = {
      indexGroup: ADSIGRP.SUMUP_READWRITE,
      indexOffset: countsymnames,
      writeBuffer: buf,
      readLength: buf.length,
      symname: 'getMultiHandle'
    }

    writeReadCommand.call(ads, commandOptions, function (err, result) {
      if (err) {
        cb.call(ads, err)
      } else {
        if (result.length > 0) {
          var resultpos = 0
          var handlespos = countsymnames*8
          handles.forEach( function (handle){
            if (typeof handle.symname !== 'undefined') {
              var adsError = result.readUInt32LE(resultpos)
              resultpos+=4
              handle.err = getError(adsError) 

              var symhandlebyte = result.readUInt32LE(resultpos)
              resultpos+=4
              if (symhandlebyte==4) {
                handle.symhandle = result.readUInt32LE(handlespos)
              }
              handlespos+=symhandlebyte 
            
              var symHandleToRelease = Buffer.alloc(4)
              symHandleToRelease.writeUInt32LE(handle.symhandle,0)
              ads.symHandlesToRelease.push(symHandleToRelease)
            }
          });
        }
        cb.call(ads, null, handles)
      }
    })
  } else cb.call(ads, null, handles)
}

**Comment by ovi1337** [comment]
Thank you for your support - maybe i miss something, but this is exactly what i already did here: https://github.com/roccomuso/node-ads/pull/12/files#diff-94814b9df475840da8a076fabd72c51dR275

I want to fetch Values with the multiRead method with the symbolName instead of the indexGroup and indexOffset. It's necessary to use the handles for fetching by name (0x0000F005 READ_/WRITE_SYMVAL_BYHANDLE). Where i'm currently stuck is the use of the handles in the multiRead:
https://github.com/roccomuso/node-ads/pull/12/files#diff-94814b9df475840da8a076fabd72c51dR381

There must be something wrong or is missing. Currently i'm using the single handle assignment for simplicity at this point. The handles are available at this point, but the usage seems to be wrong or whatever. 

__TL;DR:__
- The getHandles method is already working (just the callback is currently commented out)
- The multiRead method is working well with the indexGroup and the indexOffset.
- The multiRead method is currently unable to get Values by using symbolName(s)/Handle(s)

**Comment by ChrisHanuta** [comment]
This is the multiRead 

I can make a Merge request if you want
```
var multiRead = function (handles, cb) {
  var ads = this

  getMultiHandle.call(ads, handles, function (err, handles) {
    if (!err) {
      var countreads = 0;
      var readlen = 0;
      var bufs = null;
      handles.forEach(function(handle) {
          if (!handle.err){
            var buf = Buffer.alloc(12);
            buf.writeUInt32LE(handle.indexGroup || ADSIGRP.RW_SYMVAL_BYHANDLE,0);
            buf.writeUInt32LE(handle.symhandle,4);
            // ? if (typeof handle.arrayid !== 'undefined') {
            //  buf.writeUInt32LE(handle.totalByteLength * handle.arrayid);
            //} else {
            buf.writeUInt32LE(handle.totalByteLength,8);
            readlen += handle.totalByteLength+4;
            //}
            if (bufs) {
              bufs = Buffer.concat([bufs, buf ], bufs.length + buf.length)
            } else {  
              bufs = buf
            }
            countreads++
          }  
      })
      
      if (bufs) {
        var commandOptions = {
          indexGroup: ADSIGRP.SUMUP_READ,
          indexOffset: countreads,
          writeBuffer: bufs,
          readLength: readlen,
          symname: 'multiRead'
        }

        writeReadCommand.call(ads, commandOptions, function (err, result) {
          if (err) {
            cb.call(ads, err)
          } else {
            if (result.length > 0) {
              var resultpos = 0
              var handlespos = countreads*4
              handles.forEach( function (handle){
                if (!handle.err){
                  var adsError = result.readUInt32LE(resultpos)
                  resultpos+=4
                  if (adsError!=0) {
                    handle.err = getError(adsError) 
                  }

                  if (handle.totalByteLength>0) {
                    var integrate = Buffer.alloc(handle.totalByteLength)
                    result.copy(integrate,0,handlespos,handlespos+handle.totalByteLength);
                    integrateResultInHandle(handle, integrate)
                  }
                  handlespos+=handle.totalByteLength 
                }  
              });
            }
            cb.call(ads, null, handles)
          }
        })
      } else {
        cb.call(ads.adsClient, err)
      }
    } else {
      cb.call(ads.adsClient, err)
    }
  })
}
```

**Comment by ChrisHanuta** [comment]
I just stumbled over this line

```
      var commandOptions = {
        indexGroup: handle.indexGroup || ADSIGRP.RW_SYMVAL_BYHANDLE,
        indexOffset: handle.symhandle,
        bytelength: handle.totalByteLength,
        symname: handle.symnane
      }

      if (typeof handle.arrayid !== 'undefined') {
        commandOptions += handle.totalByteLength * handle.arrayid
      }
```
I do not understand what will be added here?

`        commandOptions += handle.totalByteLength * handle.arrayid
`

**Comment by ChrisHanuta** [comment]
If that works the handlerelease should also be switched to mulit

**Comment by ovi1337** [comment]
Yes, the auto-formatting broke it, it's currently work in progress and will be fine again before we will merge it.

**Comment by ChrisHanuta** [comment]
I think I should insert a parameter handle.arraylength. With an array, only the size needs to be increased. Then you would just have to describe the element and specify the length.

handle.arrayid comes from dumping the structures. There it make sense...

**Comment by ovi1337** [comment]
Yes, i also think we can complete get rid of `lodash`. I just did the same things like the initial stage :)
I'm also not very happy with all the names. But one step to the other.
I will do an update on my PR in some minutes.

**Comment by ovi1337** [comment]
Ah good to know about the `handle.arrayid`:)

**Comment by ovi1337** [comment]
BTW: We also should introduce data models and reorganize the whole lib.

File: README.md [hunk 1]
```diff
@@ -100,6 +100,47 @@ var client = ads.connect(options, function() {
...
+                symname: '.TESTINT',
+                indexGroup: 16448,
+                indexOffset: 26938,
+                bytelength: ads.UINT,
+            }],
+            function (err, result) {
+                if (err) {
+                    console.error(err)
+                }
+                console.log(result)
```
Comment (ChrisHanuta) [inline README.md]: we should give a hint here on the err in the results.

File: lib/ads.js [hunk 2]
```diff
@@ -510,9 +521,75 @@ var getSymbols = function (cb) {
...
+      if (result.length > 0) {
+        var resultpos = 0
+        var handlespos = commandOptions.length * 8
+
+        commandOptions.forEach(function (handle) {
+          if (handle.symname !== undefined) {
+            var adsError = result.readUInt32LE(resultpos)
+            resultpos += 4
+
+            handle.err = getError(adsError)
```
Comment (ChrisHanuta) [inline lib/ads.js]: if we export the "const errors = {" structure, we could just write the error code here. This would save memory for many errors. but we should then describe in readme.md how to evaluate it.

File: lib/ads.js [hunk 3]
```diff
@@ -266,88 +288,72 @@ var readState = function (cb) {
...
+          var handlespos = commandOptions.length * 4
+
+          commandOptions.forEach(function (handle) {
+            if (!handle.err) {
+              var adsError = result.readUInt32LE(resultpos)
+              resultpos += 4
+
+              if (adsError != 0) {
+                handle.err = getError(adsError)
+              }
```
Comment (ChrisHanuta) [inline lib/ads.js]: if we export the "const errors = {" structure, we could just write the error code here. This would save memory for many errors. but we should then describe in readme.md how to evaluate it.

File: README.md (hunk 1 above)
Comment (ChrisHanuta) [reply README.md]: console.log(result)

result.forEach(function (handle) {
  if (handle.err) {
either    console.log(handle.err)
or         console.log(ads.errors[handle.err])
  }  else {
    console.log(handle)
  }
})

File: lib/ads.js (hunk 3 above)
Comment (ChrisHanuta) [reply lib/ads.js]: handle.err = getError(adsError)

File: README.md [hunk 4]
```diff
@@ -100,6 +100,47 @@ var client = ads.connect(options, function() {
...
+            });
+```
+
+### Get handles
+
+```javascript
+        this.client.getHandles(
```
Comment (ChrisHanuta) [inline README.md]: I think the example should be like any other
```
var client = ads.connect(options, function() {
   client.getHandles(
....
}
```

File: README.md [hunk 5]
```diff
@@ -100,6 +100,47 @@ var client = ads.connect(options, function() {
 })
 ```
 
+
+### MultiRead something
+
+```javascript
+        this.client.multiRead(
```
Comment (ChrisHanuta) [inline README.md]: I think the example should be like any other

var client = ads.connect(options, function() {
   client.multiRead(
....
}

File: lib/ads.js (hunk 2 above)
Comment (ovi1337) [reply lib/ads.js]: Good point. I would like to move and clean some things today. The whole debugging parts can be also moved into another file.

File: README.md (hunk 4 above)
Comment (ovi1337) [reply README.md]: Yes i generally will update the readme today, it was very late yesterday.

Comment (ChrisHanuta) [reply README.md]: the "this.end ()" is also missing

File: README.md (hunk 5 above)
Comment (ChrisHanuta) [reply README.md]: the "this.end ()" is also missing

Comment (ChrisHanuta) [reply README.md]: excuse me sometimes meticulous

File: lib/ads.js (hunk 2 above)
Comment (ovi1337) [reply lib/ads.js]: let's do one step to the other. I think we can do that things and other cleanups and reorganizations in another PR.

**Comment by ovi1337** [comment]
bump for approvals :)

**Comment by ChrisHanuta** [comment]
@roccomuso: if the formatting is ok for you, it is for me too

**Comment by ovi1337** [comment]
@ChrisHanuta about the formatting and comparing the code i'd like to recommend you to simply switch to "split mode" if you currently don't have:
image

**Comment by ovi1337** [comment]
@ChrisHanuta fixed

**Comment by ovi1337** [comment]
@ChrisHanuta Did i something wrong? 
I absolutely appreciate your work and help and i also think that we can do it better together.
I also have the problem that @roccomuso is unable to take a bit more care to our improvement ideas and the merging of the stuff :(

**Comment by ovi1337** [comment]
I will also do my own thing now, because i'm not happy at all to spend many hours to improve this lib without any care of the owner of this fork...

**Comment by roccomuso** [comment]
Sorry, I was a little bit busy.

Let me give you access to the whole repo to be able to merge on your own
